pypeit/core/skysub.py

```diff
@@ -675,29 +675,23 @@ def local_skysub_extract(sciimg, sciivar, tilts, waveimg, global_sky, rn2_img,
     # TODO Can this be simply replaced with spat_img above (but not spat_pix since that could have holes)
     spatial_img = thismask * ximg * (np.outer(xsize, np.ones(nspat)))
 
-    # Loop over objects and group them
-    i1 = 0
-    while i1 < nobj:
-        group = np.array([], dtype=np.int)
-        group = np.append(group, i1)
+    # First, we find all groups of objects to local skysubtract together
+    groups = sobjs.get_extraction_groups()
+
+    for group in groups:
         # The default value of maskwidth = 3.0 * FWHM = 7.05 * sigma in objfind with a log(S/N) correction for bright objects
-        min_spat1 = np.maximum(sobjs[i1].TRACE_SPAT - sobjs[i1].maskwidth - 1, slit_left)
-        max_spat1 = np.minimum(sobjs[i1].TRACE_SPAT + sobjs[i1].maskwidth + 1, slit_righ)
-        for i2 in range(i1 + 1, nobj):
-            left_edge = sobjs[i2].TRACE_SPAT - sobjs[i2].maskwidth - 1
-            righ_edge = sobjs[i2].TRACE_SPAT + sobjs[i2].maskwidth + 1
-            touch = (left_edge < max_spat1) & (sobjs[i2].TRACE_SPAT > slit_left) & (righ_edge > min_spat1)
-            if touch.any():
-                max_spat1 = np.minimum(np.maximum(righ_edge, max_spat1), slit_righ)
-                min_spat1 = np.maximum(np.minimum(left_edge, min_spat1), slit_left)
-                group = np.append(group, i2)
+        left_edges = np.array([sobjs[i].TRACE_SPAT - sobjs[i].maskwidth - 1 for i in group])
+        righ_edges = np.array([sobjs[i].TRACE_SPAT + sobjs[i].maskwidth + 1 for i in group])
+
+        min_spat1 = np.maximum(np.amin(left_edges, axis=0), slit_left)
+        max_spat1 = np.minimum(np.amax(righ_edges, axis=0), slit_righ)
+
         # Create the local mask which defines the pixels that will be updated by local sky subtraction
         min_spat_img = np.outer(min_spat1, np.ones(nspat))
```

Comment:
While we are here, I noticed that this could be done more efficiently with broadcasting, which means we would not have to multiply.

Reply:
Good catch, I've replaced these (and a couple other instances) with the appropriate slices to use broadcasting. The 2 remaining `np.outer` calls in this file require more digging to convert to use broadcasting, so I'd like to punt on those.